radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubstateAccumulatorHashStore.java

```diff
@@ -67,77 +67,111 @@
 import com.radixdlt.crypto.HashUtils;
+import com.radixdlt.properties.RuntimeProperties;
 import com.radixdlt.statecomputer.forks.Forks;
 import com.radixdlt.store.DatabaseEnvironment;
 import com.radixdlt.store.berkeley.BerkeleyAdditionalStore;
 import com.radixdlt.utils.Bytes;
 import com.radixdlt.utils.Longs;
 import com.sleepycat.je.*;
+import java.io.*;
 import java.nio.charset.StandardCharsets;
 import java.util.Optional;
 import java.util.function.Function;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.bouncycastle.util.Arrays;
+import org.bouncycastle.util.encoders.Hex;
 
 public class BerkeleySubstateAccumulatorHashStore implements BerkeleyAdditionalStore {
 
   private static final Logger logger = LogManager.getLogger();
-  protected static final byte[] SUBSTATE_ACCUMULATOR_HASH_KEY =
+  private static final byte[] SUBSTATE_ACCUMULATOR_HASH_KEY =
       "substate_accumulator_hash_key".getBytes(StandardCharsets.UTF_8);
+  private static final String EPOCH_HASH_FILE_PATH =
+      "radixdlt-core/radixdlt/src/main/resources/epoch-hash";
+  public static final String EPOCH_HASH_FILE_SEPARATOR = "=";
 
   private Database substateAccumulatorHashDatabase;
   private Database epochHashDatabase;
 
   private byte[] currentSubstateAccumulatorHash;
-  private long epoch;
+  private Optional<Long> lastEpochInDbOpt;
+  private Optional<Long> lastEpochInFileOpt;
 
   private final Stopwatch watch = Stopwatch.createUnstarted();
 
   private Forks forks;
+  private final RuntimeProperties properties;
+  private Writer writer;
+
+  private boolean isUpdateEpochHashAccumulatorFileEnabled;
 
   @Inject
-  public BerkeleySubstateAccumulatorHashStore(Forks forks) {
+  public BerkeleySubstateAccumulatorHashStore(Forks forks, RuntimeProperties properties) {
     this.forks = forks;
+    this.properties = properties;
   }
 
   @Override
   public void open(DatabaseEnvironment dbEnv) {
-    this.substateAccumulatorHashDatabase =
-        dbEnv
-            .getEnvironment()
-            .openDatabase(
-                null,
-                "radix.substate_hash_accumulator",
-                new DatabaseConfig()
-                    .setAllowCreate(true)
-                    .setTransactional(true)
-                    .setKeyPrefixing(true)
-                    .setBtreeComparator(lexicographicalComparator()));
-    this.epochHashDatabase =
-        dbEnv
-            .getEnvironment()
-            .openDatabase(
-                null,
-                "radix.epoch_hash",
-                new DatabaseConfig()
-                    .setAllowCreate(true)
-                    .setTransactional(true)
-                    .setKeyPrefixing(true)
-                    .setBtreeComparator(lexicographicalComparator()));
+    this.substateAccumulatorHashDatabase = openSubstateAccumulatorHashDatabase(dbEnv);
+    this.epochHashDatabase = openEpochHashDatabase(dbEnv);
     this.currentSubstateAccumulatorHash =
```

Comment:
I think we're still missing the check we discussed yesterday - that we only allow the accumulator to start if it is actually the genesis transaction.

Also, it's more consistent with the other accumulators we use to start with a hash of 32 zeros - byte[32], not byte[0].

Reply:
The check is being done here https://radixdlt.atlassian.net/browse/NT-337.

I'll change the initial hash.